### DISCUSSION TITLE
AssertionError: IPGen failed

### DISCUSSION THREAD

**Discussion by happyxiaopeng0820** [Xilinx/finn#773] [Q&A]
![image](https://user-images.githubusercontent.com/116242566/220843374-7ea66de5-6ba5-49ab-98d0-2fa02ac04913.png)
How to deal with it?

**Comment by auphelia** [comment]
Hi @happyxiaopeng0820, have you checked the log file that is generated in the tmp folder? The path is specified in the error message.

**Reply by Ajfil** [reply]
Hi,i have same problem.this is my vitis_hls.log:
****** Vitis HLS - High-Level Synthesis from C, C++ and OpenCL v2020.2 (64-bit)
  **** SW Build 3064766 on Wed Nov 18 09:12:47 MST 2020
  **** IP Build 3064653 on Wed Nov 18 14:17:31 MST 2020
    ** Copyright 1986-2020 Xilinx, Inc. All Rights Reserved.

source /tools/Xilinx/Vitis_HLS/2020.2/scripts/vitis_hls/hls.tcl -notrace
INFO: [HLS 200-10] Running '/tools/Xilinx/Vitis_HLS/2020.2/bin/unwrapped/lnx64.o/vitis_hls'
/tools/Xilinx/Vitis_HLS/2020.2/tps/tcl/tcl8.5/tzdata/Europe/Dublin can't be opened.
INFO: [HLS 200-10] For user 'hbc' on host 'finn_dev_hbc' (Linux_x86_64 version 5.4.0-144-generic) on Fri Mar 10 14:02:37 +0000 2023
INFO: [HLS 200-10] On os Ubuntu 18.04.5 LTS
INFO: [HLS 200-10] In directory '/tmp/finn_dev_hbc/code_gen_ipgen_StreamingDataflowPartition_0_IODMA_0_eqxfbjk8'
Sourcing Tcl script '/tmp/finn_dev_hbc/code_gen_ipgen_StreamingDataflowPartition_0_IODMA_0_eqxfbjk8/hls_syn_StreamingDataflowPartition_0_IODMA_0.tcl'
HLS project: project_StreamingDataflowPartition_0_IODMA_0
HW source dir: /tmp/finn_dev_hbc/code_gen_ipgen_StreamingDataflowPartition_0_IODMA_0_eqxfbjk8
finn-hlslib dir: /home/hbc/finn/deps/finn-hlslib
custom HLS dir: /home/hbc/finn/custom_hls
INFO: [HLS 200-1510] Running: open_project project_StreamingDataflowPartition_0_IODMA_0 
INFO: [HLS 200-10] Creating and opening project '/tmp/finn_dev_hbc/code_gen_ipgen_StreamingDataflowPartition_0_IODMA_0_eqxfbjk8/project_StreamingDataflowPartition_0_IODMA_0'.
INFO: [HLS 200-1510] Running: add_files /tmp/finn_dev_hbc/code_gen_ipgen_StreamingDataflowPartition_0_IODMA_0_eqxfbjk8/top_StreamingDataflowPartition_0_IODMA_0.cpp -cflags -std=c++14 -I/home/hbc/finn/deps/finn-hlslib -I/home/hbc/finn/custom_hls 
INFO: [HLS 200-10] Adding design file '/tmp/finn_dev_hbc/code_gen_ipgen_StreamingDataflowPartition_0_IODMA_0_eqxfbjk8/top_StreamingDataflowPartition_0_IODMA_0.cpp' to the project
INFO: [HLS 200-1510] Running: set_top StreamingDataflowPartition_0_IODMA_0 
INFO: [HLS 200-1510] Running: open_solution sol1 
INFO: [HLS 200-10] Creating and opening solution '/tmp/finn_dev_hbc/code_gen_ipgen_StreamingDataflowPartition_0_IODMA_0_eqxfbjk8/project_StreamingDataflowPartition_0_IODMA_0/sol1'.
INFO: [HLS 200-1505] Using default flow_target 'vivado'
Resolution: For help on HLS 200-1505 see www.xilinx.com/cgi-bin/docs/rdoc?v=2020.2;t=hls+guidance;d=200-1505.html
INFO: [HLS 200-1510] Running: set_part xc7z020clg400-1 
INFO: [HLS 200-10] Setting target device to 'xc7z020-clg400-1'
INFO: [HLS 200-1510] Running: config_compile -disable_unroll_code_size_check -pipeline_style flp 
ERROR: [HLS 200-101] 'config_compile': Option '-pipeline_style' cannot have value 'flp'.
SYNTAX 
  config_compile [OPTIONS]
    -enable_auto_rewind[=false|*true*]
    -name_max_length <uint:*80*> 
    -no_signed_zeros[=false|*true*]
    -pipeline_loops <uint:*64*> 
    -pipeline_style <frp | *stp*>
    -pragma_strict_mode[=false|*true*]
    -unsafe_math_optimizations[=false|*true*]

SEE ALSO
  www.xilinx.com/cgi-bin/docs/rdoc?v=2020.2;t=vitis+doc;d=vyw1583260160301.html

command 'config_compile' returned error code
    while executing
"config_compile -disable_unroll_code_size_check -pipeline_style flp"
    (file "/tmp/finn_dev_hbc/code_gen_ipgen_StreamingDataflowPartition_0_IODMA_0_eqxfbjk8/hls_syn_StreamingDataflowPartition_0_IODMA_0.tcl" line 22)
    invoked from within
"source /tmp/finn_dev_hbc/code_gen_ipgen_StreamingDataflowPartition_0_IODMA_0_eqxfbjk8/hls_syn_StreamingDataflowPartition_0_IODMA_0.tcl"
    invoked from within
"hls::main /tmp/finn_dev_hbc/code_gen_ipgen_StreamingDataflowPartition_0_IODMA_0_eqxfbjk8/hls_syn_StreamingDataflowPartition_0_IODMA_0.tcl"
    ("uplevel" body line 1)
    invoked from within
"uplevel 1 hls::main {*}$newargs"
    (procedure "hls_proc" line 16)
    invoked from within
"hls_proc [info nameofexecutable] $argv"
but i don't know how to solve it
there some warnning:
there is a warnning when i run ./run-docker.sh 
![image](https://user-images.githubusercontent.com/91005367/224340679-aa4dd087-1653-4c8c-b0b3-ee10151ff22d.png)
but i already modify .bashrc
![image](https://user-images.githubusercontent.com/91005367/224342061-462f31b6-d84a-404b-8dcf-1dd1b9f33086.png)
there are some warnnig when i run cnv_end2end_example
![image](https://user-images.githubusercontent.com/91005367/224342605-1f6f5733-41f4-4f21-b82c-0ad7bad95b17.png)
![image](https://user-images.githubusercontent.com/91005367/224342771-b6a98d91-364a-4001-bf70-e1c5543ef15e.png)
![image](https://user-images.githubusercontent.com/91005367/224342936-5948be46-36d1-4867-b5fc-efc10143ddbc.png)


I'm not sure if the these warnings caused this error,i think maybe i don't run run-docker.sh in docker without root,but i'm not sure,
![image](https://user-images.githubusercontent.com/91005367/224346924-84429a93-f54d-4406-b412-97c0a628f284.png)

I ran these commands,but i see docker have Rootless mode,in this model ,i run run-docker.sh,there is warnning “WARNING: Running in rootless-mode without cgroups. To enable cgroups in rootless-mode, you need to boot the system in cgroup v2 mode."because i never use docker so i'm not sure which way is right.

**Reply by auphelia** [reply]
Hi @Ajfil, I see that your using Vivado/Vitis 2020.2. The current FINN version (v0.9) is running with tool versions 2022.1.
The error you're seeing is most likely because of changes in HLS commands, I would recommend upgrading your tools.

**Reply by Ajfil** [reply]
all right ,i will try,thank your help

**Reply by Ajfil** [reply]
hi, @auphelia ,I upgrading my Vivado,it was normal at first,but there is a new mistake
![image](https://user-images.githubusercontent.com/91005367/224486439-c1b046a8-39cf-4662-bdec-1e0e2ef27dd7.png)

here is my vitis.log :/tools/Xilinx/Vitis_HLS/2022.1/bin/loader: line 312:  7605 Killed                  "$RDI_PROG" "$@"
it seem the process is killed,is my computer problem?

**Reply by fpjentzsch** [reply]
Hi, a crash like this could happen if the tool runs out of memory. Could you keep an eye on the RAM usage while Vitis/Vivado is running?

**Reply by SAYANTANDUTTA1993** [reply]
I am facing the same issue. I have the correct Vivado and Vitis versions (2022.2). I am using AlexNet, and my system has 64 GB of RAM. On 12th step,, my memory is getting filled. Is this normal or any other issues / memory?
![SAVE_20240629_245258](https://github.com/Xilinx/finn/assets/42701167/14f6ab0b-6b48-4292-ad87-a66d900d794f)

**Reply by fpjentzsch** [reply]
If Vitis HLS is indeed running out of RAM, you could try reducing the NUM_DEFAULT_WORKERS environment variable to 1, so that FINN runs fewer Vitis HLS instances in parallel.

**Reply by SAYANTANDUTTA1993** [reply]
I tried the NUM_DEFAULT_WORKERS to 1 (shown in third picture). Except that I made some changes (shown fourth picture). But got the same error (picture 2 shows ip_gen error and picture 1 shows memory full). What is the problem then
![4](https://github.com/Xilinx/finn/assets/42701167/a7837b2d-bdc1-4bd6-8570-12963c75c7f8)
![3](https://github.com/Xilinx/finn/assets/42701167/c4606c14-5419-4a5b-9286-eb0491252829)
![2](https://github.com/Xilinx/finn/assets/42701167/ddfc0d5c-2118-4ff9-934d-983803fe4917)
![1](https://github.com/Xilinx/finn/assets/42701167/c6ee287c-1b6f-42c3-bfbe-98d5e44c81a8)
!

**Reply by fpjentzsch** [reply]
Maybe your model is simply too large for HLS to handle. You could try a smaller variant of your model, reduce the SIMD/PE parallelism, set mem_mode of the MVAU layers to "internal_decoupled" (which is easier on the synthesis) or switch to the new RTL back-end in the current dev branch.

**Comment by SAYANTANDUTTA1993** [comment]
I tried above mentioned SIMD/PE(SIMD=4/16 at max, and PE=1) , and MVAU layers(as internal_decoupled) (You can find it in below attached auto_folding_config.json file). Also changed NUM_DEFAULT_WORKERS=1 (as import os /n os.environ['NUM_DEFAULT_WORKERS']  = '1', inside jupyter notebook) but getting error ip_gen failed and memory full (my system ram is 64gb).  I am using Alex-net model which has 4,097,000 parameters and require approx 250MB size. So what is the system specification you are using or memory requirement to run such type of network.
I apologies if it's something on our end, but we're still running into the same ram issue. Please let us know if we're using to right configuration, as we want to rule out other possibilities before trying to upgrade our system as a last resort.
Thank you so much for your time
[auto_folding_config.json](https://github.com/user-attachments/files/16130151/auto_folding_config.json)
![20240708_161601](https://github.com/Xilinx/finn/assets/42701167/df84e140-0642-4b1f-bd81-921c84277acb)

**Reply by dvisa7178** [reply]
Hello! Did you solve the problem? I'm having the same issue when running a test that was supposed to work, and it's also failling at the 12th step.
